kroxylicious-filters/kroxylicious-record-validation/src/test/java/io/kroxylicious/proxy/filter/schema/validation/topic/PerRecordTopicValidatorTest.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.filter.schema.validation.topic;
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.kafka.common.message.ProduceRequestData.PartitionProduceData;
+import org.apache.kafka.common.message.ProduceRequestData.TopicProduceData;
+import org.apache.kafka.common.record.Record;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import io.kroxylicious.proxy.filter.schema.validation.Result;
+import io.kroxylicious.proxy.filter.schema.validation.record.RecordValidator;
+import io.kroxylicious.test.record.RecordTestUtils;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.InstanceOfAssertFactories.STRING;
+import static org.assertj.core.api.InstanceOfAssertFactories.list;
+import static org.assertj.core.api.InstanceOfAssertFactories.stream;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class PerRecordTopicValidatorTest {
+
+    @Mock
+    private RecordValidator validator;
+
+    @Test
+    void singleValidRecord() {
+        var tpd = createTopicProduceDataWithOneRecord(RecordTestUtils.record("good"));
+        var topicValidator = new PerRecordTopicValidator(validator);
+
+        when(validator.validate(any(Record.class))).thenReturn(Result.VALID);
+        var result = topicValidator.validateTopicData(tpd);
+        assertThat(result)
+                .succeedsWithin(Duration.ofSeconds(1))
+                .returns(false, TopicValidationResult::isAllPartitionsInvalid)
+                .returns(false, TopicValidationResult::isAnyPartitionInvalid)
+                .extracting(TopicValidationResult::invalidPartitions, stream(PartitionValidationResult.class))
+                .isEmpty();
+
+        verify(this.validator, times(1)).validate(any(Record.class));
+    }
+
+    @Test
+    void singleInvalidRecord() {
+        var tpd = createTopicProduceDataWithOneRecord(RecordTestUtils.record("bad"));
+        var topicValidator = new PerRecordTopicValidator(validator);
+
+        when(validator.validate(any(Record.class))).thenReturn(CompletableFuture.completedStage(new Result(false, "my bad record")));
+        var result = topicValidator.validateTopicData(tpd);
+        assertThat(result)
+                .succeedsWithin(Duration.ofSeconds(1))
+                .returns(true, TopicValidationResult::isAllPartitionsInvalid)
+                .returns(true, TopicValidationResult::isAnyPartitionInvalid);
+
+        assertThat(result)
+                .succeedsWithin(Duration.ofSeconds(1))
```

Comment:
Why is this two different assertions? the repetition of `succeedsWithin` feels a bit odd without using a new result.

Reply:
agreed.  I'm still learning the most pleasing/expressive way to use assert-j especially with futures.

I've extracted a variable to avoid the second succeedsWithin.